Reorganize and revise deployment

rozmansi opened the pull request:
- Reorganizes files to /var/lib appropriately
- Adds support for Fedora 36
- Allows using http://localhost eduVPN server
- Reuse config for next deployment

rozmansi committed:
Revise deployment

- Use similar paths as vpn-user-portal
- Add support for Fedora

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Readme: Update Azure Application permission guide

Azure AD apps may have "delegated" or "application" permissions. The former
is the correct one to choose when setting up our application.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Intermediate, revokeVpnConfigs: Revise, refactor

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Allow using http://localhost access to eduVPN server

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Upgrade deploy scripts to reuse config

Signed-off-by: Simon Rozman <simon@rozman.si>

FlorisHendriks commented on resources/deploy_intermediate_fedora.sh:
```diff
@@ -0,0 +1,114 @@
+sed -i "s/{applicationId}/${APPLICATION_ID}/" "/usr/share/vpn-provisioning/web/index.php"
+
+# update secret application token
+sed -i "s/{secretToken}/${${SECRET_TOKEN//\\/\\\\}//\//\\/}/" "/usr/share/vpn-provisioning/web/index.php"
```
Hey Simon, I am not really familiar with sed. What do these slashes and backslashes actually do?

This would also potentially give errors. POSIX sh can not handle nested parameters. I would suggest to use a temporarily value to store the parameter. An example can be found here https://www.shellcheck.net/wiki/SC2299

rozmansi replied:
This ugly slash/backslash dance escapes special characters sed chokes on:
* \ → \\\\  sed is using \ for backreferences. Should `$SECRET_TOKEN` contain substring "\1", it would confuse sed
* / → \\/  sed is using / as delimiters. Should `$SECRET_TOKEN` contain substring "/", the `s/{secretToken}/${SECRET_TOKEN}/` becomes invalid sed expression.

rozmansi replied:
I shall update scripts to stop using nested parameters. No problem.
Thou, do we need POSIX compliancy, being Debian and Fedora-centric for the time being? Better keep close to POSIX, I suppose.

FlorisHendriks commented on resources/deploy_intermediate_fedora.sh:
```diff
@@ -0,0 +1,114 @@
+sed -i "s/{adminApiToken}/${ADMIN_API_TOKEN}/" "/etc/vpn-provisioning"
+sed -i "s/{tenantId}/${TENANT_ID}/" "/etc/vpn-provisioning"
+sed -i "s/{applicationId}/${APPLICATION_ID}/" "/etc/vpn-provisioning"
+sed -i "s/{secretToken}/${${SECRET_TOKEN//\\/\\\\}//\//\\/}/" "/etc/vpn-provisioning"
```
Here is a nested parameter as well, please fix according to https://www.shellcheck.net/wiki/SC2299

FlorisHendriks commented on resources/deploy_intermediate_debian.sh:
```diff
@@ -0,0 +1,108 @@
+sed -i "s/{adminApiToken}/${ADMIN_API_TOKEN}/" "/etc/vpn-provisioning"
+sed -i "s/{tenantId}/${TENANT_ID}/" "/etc/vpn-provisioning"
+sed -i "s/{applicationId}/${APPLICATION_ID}/" "/etc/vpn-provisioning"
+sed -i "s/{secretToken}/${${SECRET_TOKEN//\\/\\\\}//\//\\/}/" "/etc/vpn-provisioning"
```
Here is a nested parameter as well, please fix according to https://www.shellcheck.net/wiki/SC2299

FlorisHendriks commented on resources/deploy_intermediate_debian.sh:
```diff
@@ -0,0 +1,108 @@
+sed -i "s/{applicationId}/${APPLICATION_ID}/" "/usr/share/vpn-provisioning/web/index.php"
+
+# update secret application token
+sed -i "s/{secretToken}/${${SECRET_TOKEN//\\/\\\\}//\//\\/}/" "/usr/share/vpn-provisioning/web/index.php"
```
Here is a nested parameter as well, please fix according to https://www.shellcheck.net/wiki/SC2299

rozmansi committed:
Stop nesting variable expansion in /bin/sh scripts

It's not POSIX compliant.

Reported-by: FlorisHendriks98
Resources: https://www.shellcheck.net/wiki/SC2299
Signed-off-by: Simon Rozman <simon@rozman.si>

FlorisHendriks commented on resources/deploy_intermediate_debian.sh:
```diff
@@ -0,0 +1,111 @@
+sed -i "s/{applicationId}/${APPLICATION_ID}/" "/usr/share/vpn-provisioning/web/index.php"
+
+# update secret application token
+SECRET_TOKEN_SED=${SECRET_TOKEN//\\/\\\\}
+SECRET_TOKEN_SED=${SECRET_TOKEN_SED//\//\\/}
```
Sorry I have one more comment on this. According to https://learn.microsoft.com/en-us/microsoft-365/compliance/sit-defn-azure-ad-client-secret?view=o365-worldwide, the client secret does not contain slashes or backslashes. Should we still escape them?

rozmansi replied:
Oh, nice catch. I agree. (We're not escaping Tenant ID either being either GUID or FQDN. Nor other variables where not required.)

rozmansi committed:
Stop escaping Azure AD client secret

Accordin to Microsoft documentation, they do not contain slashes / nor
backslashes \. Less is more.

Reported-by: FlorisHendriks98
Resources: https://learn.microsoft.com/en-us/microsoft-365/compliance/sit-defn-azure-ad-client-secret?view=o365-worldwide
Signed-off-by: Simon Rozman <simon@rozman.si>

FlorisHendriks commented on resources/index.php:
```diff
@@ -155,7 +155,7 @@
 $config = curl_exec($ch);
 
-$file = "/etc/eduVpnProvisioning/localDeviceIds.txt";
+$file = "/var/lib/vpn-provisioning/localDeviceIds.txt";
```
Hey Simon, I was testing the revocation of VPN configs on Fedora and I saw that the file localDeviceIds.txt was empty. It seems that Apache can not write the managedId to the localDeviceIds.txt file. It has something to do with the permissions because after following the commands described here it works for my case (https://www.linode.com/community/questions/22429/what-prevents-writing-a-file-with-correct-permissions-and-ownership)

rozmansi replied:
Sure, haven't tested this part yet. Will fix permission. Stay tuned.

rozmansi committed:
Set proper Content-Type header when returning VPN profile

The client management script should decide between OpenVPN/WireGuard based
on Content-Type of returned VPN profile, rather than profile content. But
first, the intermediate server must set the Content-Type header properly.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Fix /var/lib/vpn-provisioning permissions and SELinux context

Reported-by: FlorisHendriks98
Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Readme: Update

Signed-off-by: Simon Rozman <simon@rozman.si>

FlorisHendriks commented on resources/index.php:
```diff
@@ -152,6 +152,15 @@
 $headers[] = 'Authorization: Bearer {adminApiToken}';
 $headers[] = 'Content-Type: application/x-www-form-urlencoded';
 curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
+curl_setopt($ch, CURLOPT_HEADERFUNCTION,
```
Hey Simon, altering the Content-Type header overwrites the default header which is:

> text/html; charset=UTF-8

So if you overwrite the content-type header, PowerShell does not know how to interpret the bytes it receives from the HTTP response. As a result, we receive ascii integers as response to our invoke-webrequest:
![image](https://user-images.githubusercontent.com/47246332/213874168-ad180684-b626-4c11-85e3-f364339d732c.png)

In order to fix it I suggest we add text/html; charset=UTF-8 to the content-type header so that we receive the config in normal utf-8 text.

rozmansi replied:
The default text/html is incorrect, as the content returned by the intermediate server is _not_ HTML. The PowerShell script can be changed to interpret `application/x-wireguard-profile` and `application/x-openvpn-profile` correctly. Here's the snippet from my PowerShell that works fine with the proper Content-Type header:

```powershell
    if ($Response.StatusCode -eq 200) {
        # Using winget as system user is quite a hassle (https://github.com/microsoft/winget-cli/issues/548).
        # It can not find the winget path by itself so we need to resolve the path
        # (https://call4cloud.nl/2021/05/cloudy-with-a-chance-of-winget/#part3):

        # $ResolveWingetPath = Resolve-Path "C:\Program Files\WindowsApps\Microsoft.DesktopAppInstaller_*_x64__8wekyb3d8bbwe"
        # if ($ResolveWingetPath) {
        #     $WingetPath = $ResolveWingetPath[-1].Path
        # }
        # else {
        #     Throw "Winget not found"
        # }
        # cd $WingetPath

        # Install and deploy WireGuard tunnel if we received a wireguard-configuration
        switch ($Response.Headers['Content-Type']) {
            'application/x-wireguard-profile' {
                # .\winget.exe install --silent WireGuard.WireGuard --accept-package-agreements --accept-source-agreements | Out-Null

                # We create a new folder in the WireGuard directory.
                # We cant put it in WireGuard\Data directory as that folder is created only when we start the WireGuard application
                # (https://www.reddit.com/r/WireGuard/comments/x6f1gl/missing_data_directory_when_installing_wireguard/)
                New-Item -Path 'C:\Program Files\WireGuard' -Name 'eduVpnProvisioning' -ItemType 'directory' -ErrorAction Ignore

                # Limit access to the System user and administrators.
                icacls 'C:\Program Files\WireGuard\eduVpnProvisioning' /inheritance:r /grant:r 'SYSTEM:(OI)(CI)F' /grant:r 'Administrators:(OI)(CI)F'

                $Utf8NoBomEncoding = New-Object System.Text.UTF8Encoding $False
                [System.IO.File]::WriteAllLines('C:\Program Files\WireGuard\eduVpnProvisioning\wg0.conf', "$Response", $Utf8NoBomEncoding)
                & 'C:\Program Files\WireGuard\wireguard.exe' /installtunnelservice 'C:\Program Files\WireGuard\eduVpnProvisioning\wg0.conf'
            }
            'application/x-openvpn-profile' {
                # .\winget.exe install OpenVPNTechnologies.OpenVPN --silent --accept-package-agreements --accept-source-agreements --override "ADDLOCAL=OpenVPN.Service,OpenVPN,Drivers.TAPWindows6,Drivers" | Out-Null
                icacls "C:\Program Files\OpenVPN\config-auto" /inheritance:r /grant:r "SYSTEM:(OI)(CI)F" /grant:r "Administrators:(OI)(CI)F"
                $Utf8NoBomEncoding = New-Object System.Text.UTF8Encoding $False
                [System.IO.File]::WriteAllLines('C:\Program Files\OpenVPN\config-auto\openvpn.ovpn', "$Response", $Utf8NoBomEncoding)
                Restart-Service -Name OpenVPNService
            }
            default {
                Throw "Unexpected response: $($Response.Headers['Content-Type'])`r`n$Response"
            }
        }
```

Hadn't had time to finish reworking of the script generation (remember our talk on Signal the other day). Hence, this has not been added to this PR yet.

Nevertheless, I really suggest we send correct Content-Type header and fix the problem on the receiver side: namely PowerShell.

FlorisHendriks replied:
Alright if that is the convention then lets do that 👍

rozmansi committed:
Make Intune management script generation web-based

This simplifies script generation by dropping the requirement of a Windows
computer with Git installed.

Also, the Windows_Intune_management_script.ps1 was upgraded to work with
recent changes in intermediate server profile delivery.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Unify identation

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Forward profile's Expires HTTP header

At one point, we will have to address profile expiration and renewal on the
client side.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Make Intune management script generation website user-friendlier

As fellow Germans would say: "Mach die
Intunemanagementskriptegenerationwebsite benutzerfreundlich!" 😉

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Make Windows_Intune_management_script.ps1 repeatable

This allows admins to redeploy same script again and make devices refresh
their setup.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Revise filesystem and interface names on Windows

`eduVpnProvisioning` uses wrong letter case and renamed it to match other
parts of eduVPN server and intermediate server paths.

`wg0` is Unixism. But more importantly, it doesn't explain sysadmins where a
mysterious `wg0` network adapter came from. Therefore, renamed to `eduVPN`.
This name will also not colide with eduVPN Client for Windows: that client
sets network adapter names using VPN server FQDN.

OpenVPN does not create network adapters, so doesn't the .ovpn filename play
any significant role in our use-case. However, naming it `openvpn` might
again cause sysadmins some head-scratching where the `C:\Program
Files\OpenVPN\log\openvpn.log` came from.

If we did this change after having this in production, we should add some
cleanup code to remove previous files and services. Given this project is
still in development, the cleanup was ommited for simplicity.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Cleanup

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Simplify management script filenames

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Stabilize winget install

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Introduce uninstall management script

Should organization admins require to remove the VPN access for (some)
devices, deploying this script allows them to do so.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Fix client-cert and non-client-cert requests on same VirtualHost

Using `SSLVerifyClient require` doesn't play well with macOS curl.

Suggested-by: Floris Hendriks
Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Revise sub-CA list and add certificate details

Windows and macOS devices receive client certificates issued by "Microsoft
Intune MDM Device CA".  The "Microsoft Intune MDM Agent CA" remains allowed
for backward compatibility.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Discontinue OpenVPN support

Signed-off-by: Simon Rozman <simon@rozman.si>

FlorisHendriks commented on resources/web/management_script/index.php:
```diff
@@ -177,17 +177,12 @@
 # We start a subprocess so that we can properly log the output
 (
-id=$(security find-certificate -a | awk -F= '/issu/ && /MICROSOFT INTUNE MDM AGENT CA/ { getline; print $2 }')
+id=$(security find-certificate -a | awk -F= '/issu/ && /MICROSOFT INTUNE MDM DEVICE CA/ { getline; print $2 }')
```
Hey Simon, have you verified that this works?

When I test this (on macOS Monterey) I receive a prompt:
![image](https://user-images.githubusercontent.com/47246332/220151805-ff6d8b02-969e-42a6-9f8e-36f1c1d597ef.png)

This is due to the fact that not all applications are allowed to use the key of the MDM device cert:
![image](https://user-images.githubusercontent.com/47246332/220152150-1cf2f9ec-a675-4f55-a6fe-d27987a24c43.png)

The MDM agent certificate is created when a script gets pushed from Intune to the macOS device. Access to this key is not restricted:
![image](https://user-images.githubusercontent.com/47246332/220152433-0b86a3e7-f09b-4f6b-b69c-ac03076da8b9.png)

And will therefore not prompt the user with credentials.

rozmansi replied:
Oh dear!!! After enrolling a macOS device, I was looking for the MDM Agent certificate everywhere to no avail. Just the device certificate was there. That's why I switched to the MDM Device certificate. And I did notice the admin prompt.
Sure, the MDM Agent certificate is not only much better choice, but the only viable choice then.

rozmansi committed:
install_vpn_tunnel.sh: Upgrade

1. Move temporary MacPorts compilation directory to /var/tmp. /etc is for
config, not temp on *nix systems. Also, make the cleanup more thorough: the
tar.gz was left behind.

2. Install to /opt/vpn-provisioning. This gathers almost all our files in a
single location rather than scattering it all over /opt. Makes uninstall
easier.

3. Stop logging response to a log file. It contains WireGuard config
including private key.

4. Install Command Line Tools for Xcode if not already installed.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
uninstall_vpn_tunnel.sh: Initial version

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
install_vpn_tunnel.sh: Clean whitespace

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
(Un)Install-VPN-Tunnel.ps1: Rethrow exception

This makes Intune agent aware script failed. Without this, the exception
remain internal to the script.

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
install_vpn_tunnel.sh: Fix launchd daemon

Signed-off-by: Simon Rozman <simon@rozman.si>

rozmansi committed:
Accept GET parameters for profile retrieval too

Roopesh had a challenge POST-ing from the Swift code.

Signed-off-by: Simon Rozman <simon@rozman.si>
